### PR TITLE
Enabling matching on a subset of events

### PR DESCRIPTION
**Pull request by DonAurelio**
We enabled the reuse of previously matched events in further defined event-matching expressions. This essentially allows...
1. Separations of concerts between what we want to match on events and what we need to extract.
2. Advanced events matching and filtering using set operations natively supported in ruby.

### CODE REVIEW

**Comment by DonAurelio** [comment]
Ready for your review, Thomas.

File: bin/metababel [hunk 1]
```diff
@@ -121,46 +161,20 @@ def wrote_event_dispatchers(folder, trace_model, trace_matching_model)
...
-    dispatch_type_args += signatures.pop
-    dispatch_type = DispatchType.new(name_sanitized, dispatch_type_args, "matching_#{name_sanitized}",
-                                     matched_dispatchers)
+  automatic_dispatch_types = dispatchers.flat_map(&:dispatch_types)
 
-    matched_dispatchers.each do |dispatcher|
-      dispatcher.dispatch_types << dispatch_type
-    end
+  context = { "all" => dispatchers.to_set }
+  dispatch_types = trace_matching_model.map_event_classes_with_index do |em, _, _, default_clock_class|
```
Comment (TApplencourt) [inline bin/metababel:167]: maybe filter_map  and return either `false` or distpach type `? Like we do in the `match?`

File: lib/metababel/bt2_matching_utils.rb [hunk 2]
```diff
@@ -65,8 +66,7 @@ def match?(match_obj)
         return false
       end
 
-      # Filter one we need to extract
-      args_matched.zip(match_members).filter_map { |obj, match_obj| obj if match_obj.extract }
+      args_matched
     end
```
Comment (TApplencourt) [inline lib/metababel/bt2_matching_utils.rb:70]: Maybe put the `args_matched` in a final ese branch? Not that important neither!

**Review by TApplencourt — APPROVED** [review]
Look super nice!

File: bin/metababel [hunk 3]
```diff
@@ -121,46 +161,20 @@ def wrote_event_dispatchers(folder, trace_model, trace_matching_model)
...
 
-    dispatch_type
-  end
+    # We mutate matched_dispatchers, context.
+    get_dispatch_type(em, default_clock_class, matched_dispatchers, context)
+  end.compact
 
-  d = { dispatchers: dispatchers, dispatch_types: dispatch_types, event_name: event_name }
+  d = { dispatchers: dispatchers,
+        dispatch_types: automatic_dispatch_types + dispatch_types,
```
Comment (TApplencourt) [inline bin/metababel:176]: Maybe rename `dispatch_types` to `matching_dispatch_types` ? Just to be a little clearer

**Comment by TApplencourt** [comment]
Let's GOOOOOOOO

**Pull request merged**